File: compile-translations.py
#!/usr/bin/env python3
"""
Simple script to compile .po files to .mo files
Run this from command line: python compile-translations.py
"""

import struct
import array
import os
import re

def parse_po_file(po_file):
    """Parse a PO file and return dictionary of msgid -> msgstr"""
    with open(po_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    entries = {}
    
    # Find all msgid/msgstr pairs
    pattern = r'msgid\s+"([^"]*)"\s+msgstr\s+"([^"]*)"'
    matches = re.findall(pattern, content, re.MULTILINE)
    
    for msgid, msgstr in matches:
        # Unescape strings
        msgid = msgid.encode().decode('unicode_escape')
        msgstr = msgstr.encode().decode('unicode_escape')
        
        if msgid and msgstr:  # Skip empty strings
            entries[msgid] = msgstr
    
    return entries

def compile_mo_file(po_file, mo_file):
    """Compile a PO file to MO format"""
    entries = parse_po_file(po_file)
    
    if not entries:
        print(f"⚠ Warning: No entries found in {po_file}")
        return False
    
    # Sort by msgid
    sorted_entries = sorted(entries.items(), key=lambda x: x[0])
    
    # Generate the binary MO format
    keys = []
    values = []
    offsets = []
    
    # Build the key/value arrays
    for msgid, msgstr in sorted_entries:
        keys.append(msgid.encode('utf-8') + b'\x00')
        values.append(msgstr.encode('utf-8') + b'\x00')
    
    # Calculate offsets
    keystart = 7 * 4 + 16 * len(keys)
    valuestart = keystart + sum(len(k) for k in keys)
    koffsets = []
    voffsets = []
    
    offset = 0
    for key in keys:
        koffsets.append((len(key) - 1, keystart + offset))
        offset += len(key)
    
    offset = 0
    for value in values:
        voffsets.append((len(value) - 1, valuestart + offset))
        offset += len(value)
    
    # The header is 7 32-bit unsigned integers
    keystart = 7 * 4 + 16 * len(keys)
    valuestart = keystart + sum(len(k) for k in keys)
    koffsets = []
    voffsets = []
    
    # Calculate offsets for keys
    offset = 0
    for key in keys:
        koffsets.append((len(key) - 1, keystart + offset))
        offset += len(key)
    
    # Calculate offsets for values
    offset = 0
    for value in values:
        voffsets.append((len(value) - 1, valuestart + offset))
        offset += len(value)
    
    # Write the MO file
    with open(mo_file, 'wb') as f:
        # Magic number (little endian)
        f.write(struct.pack('I', 0x950412de))
        
        # Version
        f.write(struct.pack('I', 0))
        
        # Number of entries
        f.write(struct.pack('I', len(keys)))
        
        # Offset of table with original strings  
        f.write(struct.pack('I', 28))
        
        # Offset of table with translation strings
        f.write(struct.pack('I', 28 + len(keys) * 8))
        
        # Size of hashing table (we don't use it)
        f.write(struct.pack('I', 0))
        
        # Offset of hashing table
        f.write(struct.pack('I', 0))
        
        # Write table of original strings
        for length, offset in koffsets:
            f.write(struct.pack('I', length))
            f.write(struct.pack('I', offset))
        
        # Write table of translation strings
        for length, offset in voffsets:
            f.write(struct.pack('I', length))
            f.write(struct.pack('I', offset))
        
        # Write original strings
        for key in keys:
            f.write(key)
        
        # Write translation strings
        for value in values:
            f.write(value)
    
    return True

# Main execution
if __name__ == '__main__':
    base_dir = 'languages'
    files = [
        'packit-plugin-theme-exporter-pt_BR',
        'packit-plugin-theme-exporter-en_US'
    ]
    
    print("Compiling translation files...\n")
    
    success_count = 0
    for filename in files:
        po_file = os.path.join(base_dir, filename + '.po')
        mo_file = os.path.join(base_dir, filename + '.mo')
        
        if not os.path.exists(po_file):
            print(f"✗ Error: {po_file} not found")
            continue
        
        if compile_mo_file(po_file, mo_file):
            print(f"✓ Compiled: {po_file} -> {mo_file}")
            success_count += 1
        else:
            print(f"✗ Failed: {po_file}")
    
    print(f"\n✓ Successfully compiled {success_count}/{len(files)} translation files!")

File: compile-translations.php
<?php
/**
 * Simple script to compile .po files to .mo files
 * Run this from command line: php compile-translations.php
 */

class MO_Compiler {
    
    public function compile($po_file, $mo_file) {
        if (!file_exists($po_file)) {
            echo "Error: PO file not found: $po_file\n";
            return false;
        }
        
        $entries = $this->parse_po_file($po_file);
        $this->write_mo_file($mo_file, $entries);
        
        echo "✓ Compiled: $po_file -> $mo_file\n";
        return true;
    }
    
    private function parse_po_file($file) {
        $content = file_get_contents($file);
        $entries = array();
        
        preg_match_all('/msgid\s+"(.+?)"\s+msgstr\s+"(.+?)"/s', $content, $matches, PREG_SET_ORDER);
        
        foreach ($matches as $match) {
            $msgid = stripcslashes($match[1]);
            $msgstr = stripcslashes($match[2]);
            
            if ($msgid !== '' && $msgstr !== '') {
                $entries[$msgid] = $msgstr;
            }
        }
        
        return $entries;
    }
    
    private function write_mo_file($file, $entries) {
        $tmp = array();
        foreach ($entries as $key => $value) {
            $tmp[] = array($key, $value);
        }
        
        // Sort by original string
        usort($tmp, function($a, $b) {
            return strcmp($a[0], $b[0]);
        });
        
        // Calculate offsets
        $ids = '';
        $strings = '';
        
        foreach ($tmp as $entry) {
            $ids .= $entry[0] . "\0";
            $strings .= $entry[1] . "\0";
        }
        
        $key_offsets = array();
        $val_offsets = array();
        $offset = 0;
        
        foreach ($tmp as $entry) {
            $key_offsets[] = array($offset, strlen($entry[0]));
            $offset += strlen($entry[0]) + 1;
        }
        
        $offset = 0;
        foreach ($tmp as $entry) {
            $val_offsets[] = array($offset, strlen($entry[1]));
            $offset += strlen($entry[1]) + 1;
        }
        
        // Write MO file
        $fp = fopen($file, 'wb');
        
        // Magic number
        fwrite($fp, pack('V', 0x950412de));
        
        // File format revision
        fwrite($fp, pack('V', 0));
        
        // Number of strings
        fwrite($fp, pack('V', count($tmp)));
        
        // Offset of table with original strings
        $originals_addr = 28;
        fwrite($fp, pack('V', $originals_addr));
        
        // Offset of table with translation strings
        $translations_addr = $originals_addr + count($tmp) * 8;
        fwrite($fp, pack('V', $translations_addr));
        
        // Size of hashing table (we don't use it)
        fwrite($fp, pack('V', 0));
        
        // Offset of hashing table
        fwrite($fp, pack('V', 0));
        
        // Write original strings offsets and lengths
        $str_addr = $translations_addr + count($tmp) * 8;
        foreach ($key_offsets as $offset_data) {
            fwrite($fp, pack('V', $offset_data[1]));
            fwrite($fp, pack('V', $str_addr + $offset_data[0]));
        }
        
        // Write translation strings offsets and lengths
        $str_addr += strlen($ids);
        foreach ($val_offsets as $offset_data) {
            fwrite($fp, pack('V', $offset_data[1]));
            fwrite($fp, pack('V', $str_addr + $offset_data[0]));
        }
        
        // Write original strings
        fwrite($fp, $ids);
        
        // Write translation strings
        fwrite($fp, $strings);
        
        fclose($fp);
    }
}

// Compile both translation files
$compiler = new MO_Compiler();

$base_dir = __DIR__ . '/languages/';
$files = array(
    'packit-plugin-theme-exporter-pt_BR',
    'packit-plugin-theme-exporter-en_US'
);

echo "Compiling translation files...\n\n";

foreach ($files as $file) {
    $po_file = $base_dir . $file . '.po';
    $mo_file = $base_dir . $file . '.mo';
    $compiler->compile($po_file, $mo_file);
}

echo "\n✓ All translation files compiled successfully!\n";
